File: android/src/main/kotlin/com/google/maps/flutter/navigation/GoogleMapsAutoViewMessageHandler.kt
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.maps.flutter.navigation

import android.content.res.Resources

/** GoogleMapsAutoViewMessageHandler */
class GoogleMapsAutoViewMessageHandler(private val viewRegistry: GoogleMapsViewRegistry) :
  AutoMapViewApi {

  private fun getView(): GoogleMapsBaseMapView {
    val view = viewRegistry.getAndroidAutoView()
    if (view != null) {
      return view
    } else {
      throw FlutterError("viewNotFound", "No valid android auto view found")
    }
  }

  override fun isMyLocationEnabled(): Boolean {
    return getView().isMyLocationEnabled()
  }

  override fun setMyLocationEnabled(enabled: Boolean) {
    getView().setMyLocationEnabled(enabled)
  }

  override fun getMapType(): MapTypeDto {
    val googleMapType = getView().getMapType()
    return Convert.convertMapTypeToDto(googleMapType)
  }

  override fun setMapType(mapType: MapTypeDto) {
    val view = getView()
    val googleMapType = Convert.convertMapTypeFromDto(mapType)
    view.setMapType(googleMapType)
  }

  override fun setMapStyle(styleJson: String) {
    val view = getView()
    view.setMapStyle(styleJson)
  }

  override fun setMyLocationButtonEnabled(enabled: Boolean) {
    getView().setMyLocationButtonEnabled(enabled)
  }

  override fun setConsumeMyLocationButtonClickEventsEnabled(enabled: Boolean) {
    getView().setConsumeMyLocationButtonClickEventsEnabled(enabled)
  }

  override fun setZoomGesturesEnabled(enabled: Boolean) {
    getView().setZoomGesturesEnabled(enabled)
  }

  override fun setZoomControlsEnabled(enabled: Boolean) {
    getView().setZoomControlsEnabled(enabled)
  }

  override fun setCompassEnabled(enabled: Boolean) {
    getView().setCompassEnabled(enabled)
  }

  override fun setRotateGesturesEnabled(enabled: Boolean) {
    getView().setRotateGesturesEnabled(enabled)
  }

  override fun setScrollGesturesEnabled(enabled: Boolean) {
    getView().setScrollGesturesEnabled(enabled)
  }

  override fun setScrollGesturesDuringRotateOrZoomEnabled(enabled: Boolean) {
    getView().setScrollGesturesDuringRotateOrZoomEnabled(enabled)
  }

  override fun setTiltGesturesEnabled(enabled: Boolean) {
    getView().setTiltGesturesEnabled(enabled)
  }

  override fun setMapToolbarEnabled(enabled: Boolean) {
    getView().setMapToolbarEnabled(enabled)
  }

  override fun setTrafficEnabled(enabled: Boolean) {
    getView().setTrafficEnabled(enabled)
  }

  override fun isMyLocationButtonEnabled(): Boolean {
    return getView().isMyLocationButtonEnabled()
  }

  override fun isConsumeMyLocationButtonClickEventsEnabled(): Boolean {
    return getView().isConsumeMyLocationButtonClickEventsEnabled()
  }

  override fun isZoomGesturesEnabled(): Boolean {
    return getView().isZoomGesturesEnabled()
  }

  override fun isZoomControlsEnabled(): Boolean {
    return getView().isZoomControlsEnabled()
  }

  override fun isCompassEnabled(): Boolean {
    return getView().isCompassEnabled()
  }

  override fun isRotateGesturesEnabled(): Boolean {
    return getView().isRotateGesturesEnabled()
  }

  override fun isScrollGesturesEnabled(): Boolean {
    return getView().isScrollGesturesEnabled()
  }

  override fun isScrollGesturesEnabledDuringRotateOrZoom(): Boolean {
    return getView().isScrollGesturesEnabledDuringRotateOrZoom()
  }

  override fun isTiltGesturesEnabled(): Boolean {
    return getView().isTiltGesturesEnabled()
  }

  override fun isMapToolbarEnabled(): Boolean {
    return getView().isMapToolbarEnabled()
  }

  override fun isTrafficEnabled(): Boolean {
    return getView().isTrafficEnabled()
  }

  override fun getMyLocation(): LatLngDto? {
    val location = getView().getMyLocation() ?: return null
    return LatLngDto(location.latitude, location.longitude)
  }

  override fun getCameraPosition(): CameraPositionDto {
    return Convert.convertCameraPositionToDto(getView().getCameraPosition())
  }

  override fun getVisibleRegion(): LatLngBoundsDto {
    return Convert.convertLatLngBoundsToDto(getView().getVisibleRegion())
  }

  override fun animateCameraToCameraPosition(
    cameraPosition: CameraPositionDto,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    return getView()
      .animateCameraToCameraPosition(
        Convert.convertCameraPositionFromDto(cameraPosition),
        duration,
        callback,
      )
  }

  override fun animateCameraToLatLng(
    point: LatLngDto,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    return getView().animateCameraToLatLng(Convert.convertLatLngFromDto(point), duration, callback)
  }

  override fun animateCameraToLatLngBounds(
    bounds: LatLngBoundsDto,
    padding: Double,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    val density = Resources.getSystem().displayMetrics.density
    return getView()
      .animateCameraToLatLngBounds(
        Convert.convertLatLngBoundsFromDto(bounds),
        Convert.convertLogicalToScreenPixel(padding, density),
        duration,
        callback,
      )
  }

  override fun animateCameraToLatLngZoom(
    point: LatLngDto,
    zoom: Double,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    return getView()
      .animateCameraToLatLngZoom(Convert.convertLatLngFromDto(point), zoom, duration, callback)
  }

  override fun animateCameraByScroll(
    scrollByDx: Double,
    scrollByDy: Double,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    return getView().animateCameraByScroll(scrollByDx, scrollByDy, duration, callback)
  }

  override fun animateCameraByZoom(
    zoomBy: Double,
    focusDx: Double?,
    focusDy: Double?,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    return getView()
      .animateCameraByZoom(
        zoomBy,
        Convert.convertDeltaToPoint(focusDx, focusDy),
        duration,
        callback,
      )
  }

  override fun animateCameraToZoom(
    zoom: Double,
    duration: Long?,
    callback: (Result<Boolean>) -> Unit,
  ) {
    return getView().animateCameraToZoom(zoom, duration, callback)
  }

  override fun moveCameraToCameraPosition(cameraPosition: CameraPositionDto) {
    return getView()
      .moveCameraToCameraPosition(Convert.convertCameraPositionFromDto(cameraPosition))
  }

  override fun moveCameraToLatLng(point: LatLngDto) {
    return getView().moveCameraToLatLng(Convert.convertLatLngFromDto(point))
  }

  override fun moveCameraToLatLngBounds(bounds: LatLngBoundsDto, padding: Double) {
    val density = Resources.getSystem().displayMetrics.density
    return getView()
      .moveCameraToLatLngBounds(
        Convert.convertLatLngBoundsFromDto(bounds),
        Convert.convertLogicalToScreenPixel(padding, density),
      )
  }

  override fun moveCameraToLatLngZoom(point: LatLngDto, zoom: Double) {
    return getView().moveCameraToLatLngZoom(Convert.convertLatLngFromDto(point), zoom)
  }

  override fun moveCameraByScroll(scrollByDx: Double, scrollByDy: Double) {
    return getView().moveCameraByScroll(scrollByDx, scrollByDy)
  }

  override fun moveCameraByZoom(zoomBy: Double, focusDx: Double?, focusDy: Double?) {
    return getView().moveCameraByZoom(zoomBy, Convert.convertDeltaToPoint(focusDx, focusDy))
  }

  override fun moveCameraToZoom(zoom: Double) {
    return getView().moveCameraToZoom(zoom)
  }

  override fun followMyLocation(perspective: CameraPerspectiveDto, zoomLevel: Double?) {
    getView().followMyLocation(Convert.convertCameraPerspectiveFromDto(perspective), zoomLevel)
  }

  override fun getMinZoomPreference(): Double {
    return getView().getMinZoomPreference().toDouble()
  }

  override fun getMaxZoomPreference(): Double {
    return getView().getMaxZoomPreference().toDouble()
  }

  override fun resetMinMaxZoomPreference() {
    getView().resetMinMaxZoomPreference()
  }

  override fun setMinZoomPreference(minZoomPreference: Double) {
    getView().setMinZoomPreference(minZoomPreference.toFloat())
  }

  override fun setMaxZoomPreference(maxZoomPreference: Double) {
    getView().setMaxZoomPreference(maxZoomPreference.toFloat())
  }

  override fun getMarkers(): List<MarkerDto> {
    return getView().getMarkers()
  }

  override fun addMarkers(markers: List<MarkerDto>): List<MarkerDto> {
    return getView().addMarkers(markers)
  }

  override fun updateMarkers(markers: List<MarkerDto>): List<MarkerDto> {
    return getView().updateMarkers(markers)
  }

  override fun removeMarkers(markers: List<MarkerDto>) {
    getView().removeMarkers(markers)
  }

  override fun clearMarkers() {
    getView().clearMarkers()
  }

  override fun clear() {
    getView().clear()
  }

  override fun getPolygons(): List<PolygonDto> {
    return getView().getPolygons()
  }

  override fun addPolygons(polygons: List<PolygonDto>): List<PolygonDto> {
    return getView().addPolygons(polygons)
  }

  override fun updatePolygons(polygons: List<PolygonDto>): List<PolygonDto> {
    return getView().updatePolygons(polygons)
  }

  override fun removePolygons(polygons: List<PolygonDto>) {
    getView().removePolygons(polygons)
  }

  override fun clearPolygons() {
    getView().clearPolygons()
  }

  override fun getPolylines(): List<PolylineDto> {
    return getView().getPolylines()
  }

  override fun addPolylines(polylines: List<PolylineDto>): List<PolylineDto> {
    return getView().addPolylines(polylines)
  }

  override fun updatePolylines(polylines: List<PolylineDto>): List<PolylineDto> {
    return getView().updatePolylines(polylines)
  }

  override fun removePolylines(polylines: List<PolylineDto>) {
    getView().removePolylines(polylines)
  }

  override fun clearPolylines() {
    getView().clearPolylines()
  }

  override fun getCircles(): List<CircleDto> {
    return getView().getCircles()
  }

  override fun addCircles(circles: List<CircleDto>): List<CircleDto> {
    return getView().addCircles(circles)
  }

  override fun updateCircles(circles: List<CircleDto>): List<CircleDto> {
    return getView().updateCircles(circles)
  }

  override fun removeCircles(circles: List<CircleDto>) {
    getView().removeCircles(circles)
  }

  override fun clearCircles() {
    getView().clearCircles()
  }

  override fun registerOnCameraChangedListener() {
    getView().registerOnCameraChangedListener()
  }

  override fun isAutoScreenAvailable(): Boolean {
    return viewRegistry.getAndroidAutoView() != null
  }

  override fun setPadding(padding: MapPaddingDto) {
    getView().setPadding(padding)
  }

  override fun getPadding(): MapPaddingDto {
    return getView().getPadding()
  }
}

File: ios/google_navigation_flutter/Sources/google_navigation_flutter/GoogleMapsAutoViewMessageHandler.swift
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import GoogleMaps

enum GoogleMapsAutoViewHandlerError: Error {
  case viewNotFound
}

class GoogleMapsAutoViewMessageHandler: AutoMapViewApi {
  private let viewRegistry: GoogleMapsNavigationViewRegistry

  init(viewRegistry: GoogleMapsNavigationViewRegistry) {
    self.viewRegistry = viewRegistry
  }

  private func getView() throws -> GoogleMapsNavigationView {
    guard let view = viewRegistry.getCarPlayView() else {
      throw GoogleMapsNavigationViewHandlerError.viewNotFound
    }
    return view
  }

  func awaitMapReady(completion: @escaping (Result<Void, Error>) -> Void) {
    do {
      let mapView = try getView()
      mapView.awaitMapReady { result in
        switch result {
        case .success:
          completion(.success(()))
        case .failure(let error):
          completion(.failure(error))
        }
      }
    } catch {
      completion(.failure(error))
    }
  }

  func setMyLocationEnabled(enabled: Bool) throws {
    try getView().setMyLocationEnabled(enabled)
  }

  func getMapType() throws -> MapTypeDto {
    try Convert.convertMapType(gmsMapType: getView().getMapType())
  }

  func setMapType(mapType: MapTypeDto) throws {
    let gmsMapType = Convert.convertMapType(mapType: mapType)
    try getView().setMapType(mapType: gmsMapType)
  }

  func setMapStyle(styleJson: String) throws {
    try getView().setMapStyle(styleJson: styleJson)
  }

  func setMyLocationButtonEnabled(enabled: Bool) throws {
    try getView().setMyLocationButtonEnabled(enabled)
  }

  func setZoomGesturesEnabled(enabled: Bool) throws {
    try getView().setZoomGesturesEnabled(enabled)
  }

  func setZoomControlsEnabled(enabled: Bool) throws {
    try getView().setZoomControlsEnabled(enabled)
  }

  func setCompassEnabled(enabled: Bool) throws {
    try getView().setCompassEnabled(enabled)
  }

  func setRotateGesturesEnabled(enabled: Bool) throws {
    try getView().setRotateGesturesEnabled(enabled)
  }

  func setScrollGesturesEnabled(enabled: Bool) throws {
    try getView().setScrollGesturesEnabled(enabled)
  }

  func setScrollGesturesDuringRotateOrZoomEnabled(enabled: Bool) throws {
    try getView().setScrollGesturesDuringRotateOrZoomEnabled(enabled)
  }

  func setTiltGesturesEnabled(enabled: Bool) throws {
    try getView().setTiltGesturesEnabled(enabled)
  }

  func setMapToolbarEnabled(enabled: Bool) throws {
    try getView().setMapToolbarEnabled(enabled)
  }

  func setTrafficEnabled(enabled: Bool) throws {
    try getView().setTrafficEnabled(enabled)
  }

  func setTrafficPromptsEnabled(enabled: Bool) throws {
    try getView().setTrafficPromptsEnabled(enabled)
  }

  func setTrafficIncidentCardsEnabled(enabled: Bool) throws {
    try getView().setTrafficIncidentCardsEnabled(enabled)
  }

  func setReportIncidentButtonEnabled(enabled: Bool) throws {
    try getView().setReportIncidentButtonEnabled(enabled)
  }

  func setAutoMapOptions(mapOptions: AutoMapOptionsDto) throws {
    // Store the map options in BaseCarSceneDelegate static property
    let options = AutoMapViewOptions(
      cameraPosition: mapOptions.cameraPosition != nil
        ? Convert.convertCameraPosition(position: mapOptions.cameraPosition!) : nil,
      mapId: mapOptions.mapId,
      mapType: mapOptions.mapType != nil
        ? Convert.convertMapType(mapType: mapOptions.mapType!) : nil,
      mapColorScheme: mapOptions.mapColorScheme != nil
        ? Convert.convertMapColorScheme(mapColorScheme: mapOptions.mapColorScheme!) : nil,
      forceNightMode: mapOptions.forceNightMode != nil
        ? Convert.convertNavigationForceNightMode(forceNightMode: mapOptions.forceNightMode!) : nil
    )
    BaseCarSceneDelegate.mapOptions = options
  }

  func isMyLocationEnabled() throws -> Bool {
    try getView().isMyLocationEnabled()
  }

  func isMyLocationButtonEnabled() throws -> Bool {
    try getView().isMyLocationButtonEnabled()
  }

  func isZoomGesturesEnabled() throws -> Bool {
    try getView().isZoomGesturesEnabled()
  }

  func isZoomControlsEnabled() throws -> Bool {
    try getView().isZoomControlsEnabled()
  }

  func isCompassEnabled() throws -> Bool {
    try getView().isCompassEnabled()
  }

  func isRotateGesturesEnabled() throws -> Bool {
    try getView().isRotateGesturesEnabled()
  }

  func isScrollGesturesEnabled() throws -> Bool {
    try getView().isScrollGesturesEnabled()
  }

  func isScrollGesturesEnabledDuringRotateOrZoom() throws -> Bool {
    try getView().isScrollGesturesEnabledDuringRotateOrZoom()
  }

  func isTiltGesturesEnabled() throws -> Bool {
    try getView().isTiltGesturesEnabled()
  }

  func isMapToolbarEnabled() throws -> Bool {
    try getView().isMapToolbarEnabled()
  }

  func isTrafficEnabled() throws -> Bool {
    try getView().isTrafficEnabled()
  }

  func isTrafficPromptsEnabled() throws -> Bool {
    try getView().isTrafficPromptsEnabled()
  }

  func isTrafficIncidentCardsEnabled() throws -> Bool {
    try getView().isTrafficIncidentCardsEnabled()
  }

  func isReportIncidentButtonEnabled() throws -> Bool {
    try getView().isReportIncidentButtonEnabled()
  }

  func getMyLocation() throws -> LatLngDto? {
    do {
      guard let myLocation = try getView().getMyLocation() else {
        return nil
      }
      return LatLngDto(
        latitude: myLocation.latitude,
        longitude: myLocation.longitude
      )
    } catch {
      throw error
    }
  }

  func getCameraPosition() throws -> CameraPositionDto {
    try Convert.convertCameraPosition(position: getView().getCameraPosition())
  }

  func getVisibleRegion() throws -> LatLngBoundsDto {
    try Convert.convertLatLngBounds(bounds: getView().getVisibleRegion())
  }

  func animateCameraToCameraPosition(
    cameraPosition: CameraPositionDto,
    duration: Int64?,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    do {
      try getView()
        .animateCameraToCameraPosition(
          cameraPosition:
            Convert
            .convertCameraPosition(position: cameraPosition))

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func animateCameraToLatLng(
    point: LatLngDto, duration: Int64?,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    do {
      try getView().animateCameraToLatLng(point: Convert.convertLatLngFromDto(point: point))

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func animateCameraToLatLngBounds(
    bounds: LatLngBoundsDto,
    padding: Double, duration: Int64?,
    completion:
      @escaping (
        Result<
          Bool,
          Error
        >
      )
      -> Void
  ) {
    do {
      try getView().animateCameraToLatLngBounds(
        bounds: Convert.convertLatLngBounds(bounds: bounds),
        padding: padding
      )

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func animateCameraToLatLngZoom(
    point: LatLngDto, zoom: Double,
    duration: Int64?,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    do {
      try getView().animateCameraToLatLngZoom(
        point: Convert.convertLatLngFromDto(point: point),
        zoom: zoom
      )

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func animateCameraByScroll(
    scrollByDx: Double, scrollByDy: Double,
    duration: Int64?,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    do {
      try getView().animateCameraByScroll(dx: scrollByDx, dy: scrollByDy)

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func animateCameraByZoom(
    zoomBy: Double, focusDx: Double?,
    focusDy: Double?, duration: Int64?,
    completion:
      @escaping (
        Result<
          Bool,
          Error
        >
      ) -> Void
  ) {
    do {
      let focus = Convert.convertDeltaToPoint(dx: focusDx, dy: focusDy)
      try getView().animateCameraByZoom(zoomBy: zoomBy, focus: focus)

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func animateCameraToZoom(
    zoom: Double, duration: Int64?,
    completion: @escaping (Result<Bool, Error>) -> Void
  ) {
    do {
      try getView().animateCameraToZoom(zoom: zoom)

      // No callback supported, just return immediately
      completion(.success(true))
    } catch {
      completion(.failure(error))
    }
  }

  func moveCameraToCameraPosition(cameraPosition: CameraPositionDto) throws {
    try getView()
      .moveCameraToCameraPosition(
        cameraPosition:
          Convert
          .convertCameraPosition(position: cameraPosition))
  }

  func moveCameraToLatLng(point: LatLngDto) throws {
    try getView().moveCameraToLatLng(point: Convert.convertLatLngFromDto(point: point))
  }

  func moveCameraToLatLngBounds(
    bounds: LatLngBoundsDto,
    padding: Double
  ) throws {
    try getView().moveCameraToLatLngBounds(
      bounds: Convert.convertLatLngBounds(bounds: bounds),
      padding: padding
    )
  }

  func moveCameraToLatLngZoom(point: LatLngDto, zoom: Double) throws {
    try getView().moveCameraToLatLngZoom(
      point: Convert.convertLatLngFromDto(point: point),
      zoom: zoom
    )
  }

  func moveCameraByScroll(scrollByDx: Double, scrollByDy: Double) throws {
    try getView().moveCameraByScroll(dx: scrollByDx, dy: scrollByDy)
  }

  func moveCameraByZoom(
    zoomBy: Double, focusDx: Double?,
    focusDy: Double?
  ) throws {
    let focus = Convert.convertDeltaToPoint(dx: focusDx, dy: focusDy)
    return try getView().moveCameraByZoom(zoomBy: zoomBy, focus: focus)
  }

  func moveCameraToZoom(zoom: Double) throws {
    try getView().moveCameraToZoom(zoom: zoom)
  }

  func followMyLocation(
    perspective: CameraPerspectiveDto,
    zoomLevel: Double?
  ) throws {
    try getView()
      .followMyLocation(
        perspective: Convert.convertCameraPerspective(perspective: perspective),
        zoomLevel: zoomLevel
      )
  }

  func getMarkers() throws -> [MarkerDto] {
    try getView().getMarkers()
  }

  func addMarkers(markers: [MarkerDto]) throws -> [MarkerDto] {
    try getView().addMarkers(markers: markers)
  }

  func updateMarkers(markers: [MarkerDto]) throws -> [MarkerDto] {
    try getView().updateMarkers(markers: markers)
  }

  func removeMarkers(markers: [MarkerDto]) throws {
    try getView().removeMarkers(markers: markers)
  }

  func clearMarkers() throws {
    try getView().clearMarkers()
  }

  func getPolygons() throws -> [PolygonDto] {
    try getView().getPolygons()
  }

  func addPolygons(polygons: [PolygonDto]) throws -> [PolygonDto] {
    try getView().addPolygons(polygons: polygons)
  }

  func updatePolygons(polygons: [PolygonDto]) throws -> [PolygonDto] {
    try getView().updatePolygons(polygons: polygons)
  }

  func removePolygons(polygons: [PolygonDto]) throws {
    try getView().removePolygons(polygons: polygons)
  }

  func clearPolygons() throws {
    try getView().clearPolygons()
  }

  func getPolylines() throws -> [PolylineDto] {
    try getView().getPolylines()
  }

  func addPolylines(polylines: [PolylineDto]) throws -> [PolylineDto] {
    try getView().addPolylines(polylines: polylines)
  }

  func updatePolylines(polylines: [PolylineDto]) throws -> [PolylineDto] {
    try getView().updatePolylines(polylines: polylines)
  }

  func removePolylines(polylines: [PolylineDto]) throws {
    try getView().removePolylines(polylines: polylines)
  }

  func clearPolylines() throws {
    try getView().clearPolylines()
  }

  func getCircles() throws -> [CircleDto] {
    try getView().getCircles()
  }

  func addCircles(circles: [CircleDto]) throws -> [CircleDto] {
    try getView().addCircles(circles: circles)
  }

  func updateCircles(circles: [CircleDto]) throws -> [CircleDto] {
    try getView().updateCircles(circles: circles)
  }

  func removeCircles(circles: [CircleDto]) throws {
    try getView().removeCircles(circles: circles)
  }

  func clearCircles() throws {
    try getView().clearCircles()
  }

  func clear() throws {
    try getView().clear()
  }

  func setConsumeMyLocationButtonClickEventsEnabled(enabled: Bool) throws {
    try getView().setConsumeMyLocationButtonClickEventsEnabled(enabled: enabled)
  }

  func isConsumeMyLocationButtonClickEventsEnabled() throws -> Bool {
    try getView().isConsumeMyLocationButtonClickEventsEnabled()
  }

  func getMinZoomPreference() throws -> Double {
    try Double(getView().getMinZoomPreference())
  }

  func getMaxZoomPreference() throws -> Double {
    try Double(getView().getMaxZoomPreference())
  }

  func resetMinMaxZoomPreference() throws {
    try getView().resetMinMaxZoomPreference()
  }

  func setMinZoomPreference(minZoomPreference: Double) throws {
    try getView().setMinZoomPreference(minZoomPreference: Float(minZoomPreference))
  }

  func setMaxZoomPreference(maxZoomPreference: Double) throws {
    try getView().setMaxZoomPreference(maxZoomPreference: Float(maxZoomPreference))
  }

  func enableOnCameraChangedEvents() throws {
    try getView().enableOnCameraChangedEvents()
  }

  func isAutoScreenAvailable() throws -> Bool {
    viewRegistry.getCarPlayView() != nil
  }

  func setPadding(padding: MapPaddingDto) throws {
    try getView().setPadding(padding: padding)
  }

  func getPadding() throws -> MapPaddingDto {
    try getView().getPadding()
  }

  func getMapColorScheme() throws -> MapColorSchemeDto {
    try Convert.convertMapColorScheme(uiUserInterfaceStyle: getView().getMapColorScheme())
  }

  func setMapColorScheme(mapColorScheme: MapColorSchemeDto) throws {
    let colorScheme = Convert.convertMapColorScheme(mapColorScheme: mapColorScheme)
    try getView().setMapColorScheme(colorScheme: colorScheme)
  }

  func getForceNightMode() throws -> NavigationForceNightModeDto {
    try Convert.convertNavigationForceNightMode(
      gmsNavigationLightingMode: getView().getForceNightMode()
    )
  }

  func setForceNightMode(forceNightMode: NavigationForceNightModeDto) throws {
    let nightMode = Convert.convertNavigationForceNightMode(
      forceNightMode: forceNightMode
    )
    try getView().setForceNightMode(nightMode)
  }

  func sendCustomNavigationAutoEvent(event: String, data: Any) throws {
    // This method receives custom events from Flutter.
    // The implementation is left empty by design, as developers should handle
    // custom events in their BaseCarSceneDelegate subclass by overriding
    // onCustomNavigationAutoEventFromFlutter method.
    //
    // Note: If you need to handle events here, you would need to maintain a reference
    // to your CarSceneDelegate instance and call a method on it.
  }
}
